pyscroll/data.py

```diff
@@ -393,92 +405,127 @@ class MapAggregator(PyscrollDataAdapter):
         # so we must move all the offsets if there is a negative so
         # that all the tile coordinates are >= (0, 0)
-        self.maps.append((data, rect))
+        self.maps.append((data, rect, z_index))
+        self._adjust_map_positions(offset)
+        self._update_map_size()
+
+    def remove_map(self, data: PyscrollDataAdapter) -> None:
+        """
+        Removes a map from the aggregator.
+
+        Args:
+            data: The data adapter of the map to remove.
+        """
+        if data not in [m[0] for m in self.maps]:
+            raise ValueError("Map is not in the aggregator")
+        self.maps = [m for m in self.maps if m[0] != data]
+        self._update_map_size()
+
+    def _adjust_map_positions(self, offset: tuple[int, int]) -> None:
+        """Adjusts map positions based on negative offsets."""
+        ox = self._min_x - offset[0]
+        oy = self._min_y - offset[1]
         if ox > 0 or oy > 0:
-            for data, rect in self.maps:
+            for _, rect, _ in self.maps:
                 rect.move_ip((ox, oy))
-                mx = max(mx, rect.right)
-                my = max(my, rect.bottom)
         else:
-            rect.move_ip(-self._min_x, -self._min_y)
+            for _, rect, _ in self.maps:
+                rect.move_ip(-self._min_x, -self._min_y)
+
+    def _update_map_size(self) -> None:
+        """Updates the overall map size."""
+        mx = 0
+        my = 0
+        for _, rect, _ in self.maps:
             mx = max(mx, rect.right)
             my = max(my, rect.bottom)
         self.map_size = mx, my
 
-    def remove_map(self, data: PyscrollDataAdapter):
-        """
-        Remove map - not implemented
-
-        """
-        raise NotImplementedError
-
     def get_animations(self) -> None:
         """
         Get animations - not implemented
-
         """
-        pass
+        return None
 
     def reload_data(self) -> None:
         """
         Reload the tiles - not implemented
-
         """
-        pass
+        return None
 
     @property
-    def visible_tile_layers(self):
+    def visible_tile_layers(self) -> list[int]:
+        """
+        Returns a sorted list of all visible tile layers from all added maps.
+        """
         layers = set()
-        for data, offset in self.maps:
+        for data, _, _ in self.maps:
             layers.update(list(data.visible_tile_layers))
         return sorted(layers)
 
-    def get_tile_images_by_rect(self, view: RectLike):
+    def get_tile_images_by_rect(
+        self, view: pygame.Rect
+    ) -> Iterable[tuple[int, int, int, Surface]]:
+        """
+        Yields tile images within the given view rectangle from all added maps.
+
+        Args:
+            view: Rect-like object defining the tile view.
+        """
         view = pygame.Rect(view)
-        for data, rect in self.maps:
+        sorted_maps = sorted(self.maps, key=lambda x: x[2])  # Sort by z-index
```

Comment:
maps should be sorted when added or their z index changed.  its inefficient to sort the maps each render, since you can just put them in render order in the `maps` list